fast_llm/tools/prepare_dataset.py

```diff
@@ -0,0 +1,359 @@
+        desc="Distributed backend to use.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.incl, torch.distributed.Backend.backend_list),
+    )
+
+    def _validate(self):
+        if self.world_size is None:
+            self.world_size = self.default_world_size
+        if self.rank is None:
+            self.rank = self.default_rank
+        super()._validate()
+        Assert.in_range(self.rank, 0, self.world_size)
+
+
+@config_class()
+class DatasetPreparatorConfig(RunnableConfig):
+    _abstract = True
+    model_name: typing.ClassVar[str]
+
+    output_path: pathlib.Path = Field(
+        desc="Output directory for the processed dataset.",
+        hint=FieldHint.core,
+    )
+    distributed: DistributedConfig = Field(
+        default_factory=DistributedConfig,
+        desc="Configuration for distributed processing.",
+        hint=FieldHint.feature,
+    )
+
+    @classmethod
+    def get_dataset_preparator_class(cls) -> typing.Type["DatasetPreparator"]:
+        raise NotImplementedError
+
+    def _get_runnable(self, parsed: argparse.Namespace) -> typing.Callable[[], None]:
+        dataset_preparator = self.get_dataset_preparator_class()(config=self)
+        return dataset_preparator.run
+
+
+class DatasetPreparator(abc.ABC):
+    _abstract = True
+    _config: DatasetPreparatorConfig
+    config_class: typing.ClassVar[type[DatasetPreparatorConfig]] = DatasetPreparatorConfig
+
+    def __init__(self, config: DatasetPreparatorConfig) -> None:
+        Assert.custom(isinstance, config, self.config_class)
+        config.validate()
+        self._config = config
+
+    def run(self) -> None:
+        raise NotImplementedError
+
+
+@config_class
+class GPTDatasetConfig(Config):
+    name_or_path: str = Field(
+        desc="Name or path of the dataset.",
+        hint=FieldHint.core,
+    )
+    config_name: None | str = Field(
+        default=None,
+        desc="Specific configuration name for the dataset.",
+        hint=FieldHint.optional,
+    )
+    split: str = Field(
+        default="train",
+        desc="Split of the dataset to use.",
+        hint=FieldHint.optional,
+    )
+    field: str = Field(
+        default="text",
+        desc="Field of the dataset to use.",
+        hint=FieldHint.optional,
+    )
+    data_type: DataType = Field(
+        default=None,
+        desc="Data type of the dataset field.",
+        hint=FieldHint.derived,
+    )
+    trust_remote_code: bool = Field(
+        default=False,
+        desc="Trust remote code when downloading the dataset.",
+        hint=FieldHint.optional,
+    )
+    disable_disk_space_check: bool = Field(
+        default=False,
+        desc="Disable disk space check. Useful for environments where disk space is not accurately reported.",
+        hint=FieldHint.optional,
+    )
+
+
+@config_class()
+class GPTDatasetPreparatorConfig(DatasetPreparatorConfig):
+    _abstract = False
+    model_name: typing.ClassVar[str] = "gpt"
```

Comment:
I don't think we have a 1-1 match between datasets and models. The new dataset stuff leaves room for other GPT dataset formats, and there can be plenty of other scenarios for other models.

So I think we should decouple preparators from models. This one would have `preparator_name = "gpt_memmap"` and class `GPTMemmapDatasetPreparator[Config]`